site-src/geps/gep-713.md

```diff
@@ -0,0 +1,409 @@
+In some cases (likely limited to mesh) we may want to apply policies to requests
+to external services. To accomplish this, implementations can choose to support
+a refernce to a virtual resource type:
+
+```yaml
+kind: RetryPolicy
+apiVersion: networking.acme.io/v1alpha1
+metadata:
+  name: foo
+spec:
+  default:
+    maxRetries: 5
+  targetRef:
+    group: networking.acme.io
+    kind: ExternalService
+    hostname: foo.com
+```
+
+### Apply Policies to Specific Matches
+In some cases we may want to target specific matches within nested objects. For
+example, it may be useful to attach policies to a specific Gateway listener or
+Route rule.
+
+Each Route rule or Gateway listener should be expanded with an optional name
+field. The target ref would be expanded with an optional sectionName field that
+could be used to refer to that specific section of the resource. It would refer
+to the following concepts on these resources:
+
+* Gateway.Listeners.Name
+* xRoute.Rules.Name
+* Service.Ports.Name
+
+```yaml
+kind: HTTPRoute
+apiVersion: networking.x-k8s.io/v1alpha1
+metadata:
+  name: http-app-1
+  labels:
+    app: foo
+spec:
+  hostnames:
+  - "foo.com"
+  rules:
+  - name: bar
+    matches:
+    - path:
+        type: Prefix
+        value: /bar
+    forwardTo:
+    - serviceName: my-service1
+      port: 8080
+---
+kind: RetryPolicy
+apiVersion: networking.acme.io/v1alpha1
+metadata:
+  name: foo
+spec:
+  maxRetries: 5
+  targetRef:
+    name: foo
+    group: networking.x-k8s.io
+    kind: HTTPRoute
+    sectionName: bar
+```
+
+This would require adding a `Name` field to Gateway listeners and Route rules:
+
+```go
+type Listener struct {
+    // Name is the name of the Listener.
+    //
+    // Support: Core
+    //
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=253
+    // +optional
+    Name string `json:"name,omitempty"`
+    // ...
+}
+```
+
+```go
+type RouteRule struct {
+    // Name is the name of the Route rule.
+    //
+    // Support: Core
+    //
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=253
+    // +optional
+    Name string `json:"name,omitempty"`
+    // ...
+}
+```
+
+### Kubectl Plugin
+To help improve UX and standardization, a kubectl plugin will be developed that
+will be capable of describing the computed sum of policy that applies to a given
+resource, including policies applied to parent resources.
+
+Each Policy CRD that wants to be supported by this plugin will need to follow
+the API structure defined above and add a `networking.x-k8s.io/policy: true` to
+the CRD.
+
+### Interaction with Custom Filters
+The concept of policy attachment is rather similar to attaching custom filters
+to a Route rule. Although it's possible that both concepts could be supported
+concurrently in the future, it is strongly recommended that implementations not
+recycle the same resource types for both concepts. To help simplify
```

Comment:
I would suggest making this paragraph more generic:

```
### Guidance for implementors on when and how to use policy attachment
There are multiple methods of custom extension in the Gateway API. Policy attachment and custom Route filters are two of these. Policy attachment is designed to provide arbitrary configuration fields that decorate Gateway API resources. Route filters provide custom request/response filters embedded inside Route resources. Both are extension methods for fields that cannot easily be standardized as core or extended fields of the Gateway API. The following guidance should be considered when introducing a custom field into any Gateway controller implementation:

1) For any given field that a Gateway controller implementation needs, the possibility of using core or extended should always be considered before using custom policy resources. This is encouraged to promote standardization and, over time, to absorb capabilities into the API as 1st class fields, which offer a more streamlined UX than custom policy attachment. 

2) Although it's possible that arbitrary fields could be supported by custom policy, custom route filters, and core/extended fields concurrently, it is strongly recommended that implementations not use multiple mechanisms for representing the same fields. A given field should only be supported through a single extension method. An example of potential conflict is policy precedence and structured hierarchy, which only applies to custom policies. Allowing a field to exist in custom policies and also other areas of the API, which are not part of the structured hierarchy, breaks the precedence model. Note that this guidance may change in the future as we gain a better understanding for extension mechanisms of the Gateway API can interoperate.
```

Reply:
Also, it doesn't have to be in this GEP, but in our docs we should probably have a more detailed picture of how core, extended, and custom portions of the API relate to each other. "Custom" is more of a concept at this point than a tangible thing as custom is really two different things - custom policy attachment and route fillters. Both are two very different custom extension methods, but we don't currently make it very clear in our docs that these two things are what comprise our "custom API".

![image](https://user-images.githubusercontent.com/17914385/125455369-6e1974bc-246f-46cd-b9d4-abada3f04a3d.png)

Reply:
Thanks for the great content! I've added your suggestions to the GEP. Agree that we should include a better description of the distinctions between custom, extended, and core but that it can be added separately.